SEP-2633: Standard Client-Side Configuration Format - mcp.json (#2633)
modelcontextprotocol/modelcontextprotocol: docs/seps/2633-standard-client-mcp-json-configuration.mdx

| **Title**     | Standard Client-Side Configuration Format - mcp.json                                                                            |
| **Status**    | Draft                                                                                                                           |
| **Type**      | Standards Track                                                                                                                 |
| **Created**   | 2026-04-21                                                                                                                      |
| **Author(s)** | Bob Dickinson ([@BobDickinson](https://github.com/BobDickinson)), Tadas Antanavicius ([@tadasant](https://github.com/tadasant)) |
| **Sponsor**   | Tadas Antanavicius ([@tadasant](https://github.com/tadasant))                                                                   |
| **PR**        | [#2633](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2633)                                                 |

---

## Abstract

> The MCP ecosystem has server.json (the MCP Registry package specification), and soon the related ServerCard, for describing how servers can be configured. But there's no standard for the other side: how a client will connect to its servers.

Reviewer: ```suggestion
The MCP ecosystem has [server.json](https://github.com/modelcontextprotocol/registry/blob/main/docs/reference/server-json/generic-server-json.md) (the MCP Registry package specification), and soon the related [Server Card](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2127), for describing how servers can be configured. But there's no standard for the other side: how a client will connect to its servers.
```

## Specification

We will establish a standard format for mcp.json as defined below.

### Relationship to server.json

| Format      | Purpose                                     | Configurability                                                        |
| :---------- | :------------------------------------------ | :--------------------------------------------------------------------- |
| server.json | Server package specification for registries | Highly configurable — variables, templates, user-adjustable parameters |
| mcp.json    | Client-side server configuration            | Fully resolved — only auth secrets remain as interpolatable variables  |

- server.json = "Here's how this server can be configured"
- mcp.json = "Here's exactly how this client will connect to its servers"

Each entry in an mcp.json is what you get after resolving a server.json template with concrete values.

### File Name, Type, and Structure

The file will be named `mcp.json` and will be a JSON file (and explicitly not a JSONC file). MCP servers will be under the top-level key `mcpServers`, and will consist of a map of server names to server configurations:

```json
{

>   "mcpServers": {

Reviewer: Do we definitely want/need to define `mcpServers` like this?

Could we potentially scope this spec down to be "the object you put _inside_ `mcpServers`"; implying the choice of e.g. `mcpServers` vs. `servers` is up to the client?

Maybe that's confusing, because then we're not really replacing what is commonly today called the `.mcp.json` file, we're more just replacing the object in the `mcpServers` field... but that also seems potentially cleaner, in that we're not trying to wholesale replace that whole file (what with e.g. VS Code's management of `inputs`, etc).

And maybe the design goal here _is_ to define the whole file, so you can copy/paste the whole thing... probably don't feel strongly on this, but what do you think?

    "server-name": { ... },
    "another-server": { ... }
  }
}
```

> Server names must match `^[a-zA-Z0-9_\[\]-]+$` (alphanumeric, hyphens, underscores, brackets).

Reviewer: We should probably allow for `name` as defined in server.json, i.e. allow `.` and `/` as part of DNS prefixing. But of course this is a "localized" name that doesn't _have_ to align with `name`.

#### Transport-Specific Requirements

**stdio servers** (`type: "stdio"`):

- `command` is required
- `args` is allowed
- `env` is allowed
- `url` and `headers` are not allowed

**Remote servers** (`type: "sse"` or `type: "streamable-http"`):

- `url` is required
- `headers` is allowed
- `oauth` is allowed (see [OAuth Configuration](#oauth-configuration))
- `command`, `args`, and `env` are not allowed

### Transport Types

#### stdio — Local Process

Runs a local process that communicates via stdin/stdout:

```json
{
  "github": {
    "title": "GitHub",
    "description": "Create and manage issues, PRs, branches, and files.",
    "type": "stdio",
    "command": "npx",
    "args": ["-y", "@modelcontextprotocol/server-github@0.6.2"],
    "env": {
      "GITHUB_PERSONAL_ACCESS_TOKEN": "${env:GITHUB_PERSONAL_ACCESS_TOKEN}"
    }
  }
}
```

#### sse — Server-Sent Events

For remote servers using the SSE transport:

```json
{
  "monitoring": {
    "title": "Monitoring",
    "type": "sse",
    "url": "https://mcp.monitoring.example.com/sse",
    "headers": {
      "Authorization": "Bearer ${env:MONITORING_TOKEN}"
    }
  }
}
```

> #### streamable-http — HTTP Streaming

Reviewer: ```suggestion
#### streamable-http — Streamable HTTP
```

> For remote servers using HTTP streaming:

Reviewer: ```suggestion
For remote servers using streamable HTTP:
```

```json
{
  "analytics": {
    "title": "Analytics",
    "type": "streamable-http",
    "url": "https://mcp.analytics.example.com/mcp",
    "headers": {
      "X-API-Key": "${env:ANALYTICS_API_KEY}"
    }
  }
}
```

> ### OAuth Configuration

Reviewer: I [recently added](https://github.com/pulsemcp/air/blob/main/docs/mcp-json-proposal.md#oauth-configuration) a couple OAuth fields I think are necessary in some cases / matches what [Claude Code offers](https://code.claude.com/docs/en/mcp#override-oauth-metadata-discovery) to my implementation.

IIRC these additional fields are needed for the official BigQuery MCP server, as an example.
